src/lazy/binary/immutable_buffer.rs

```diff
@@ -157,6 +159,257 @@ impl<'a> ImmutableBuffer<'a> {
         }
     }
 
+    /// Reads a [`FlexInt`] from the buffer.
+    pub fn read_flex_int(self) -> ParseResult<'a, FlexInt> {
+        // A FlexInt has the same structure as a FlexUInt. We can read a FlexUInt and then re-interpret
+        // its unsigned bytes as two's complement bytes.
+        let (flex_uint, remaining) = self.read_flex_primitive_as_uint("reading a FlexInt", true)?;
+        let unsigned_value = flex_uint.value();
+
+        // If the encoded FlexInt required `N` bytes to encode where `N` is fewer than 8, then its
+        // u64 value will have `8 - N` leading zero bytes. If the highest bit in the encoding was a
+        // 1, then then number is negative and we need to flip all of those leading zeros to ones.
+        // Look at the original input to see if the highest bit was a zero (positive) or one (negative).
+        let last_explicit_byte = self.bytes()[flex_uint.size_in_bytes() - 1];
+        let sign_bit = 0b1000_0000 & last_explicit_byte;
+        let signed_value = if sign_bit == 0 {
+            unsigned_value as i64
+        } else {
+            // Flip all of the leading zeros in the unsigned value to ones and then re-interpret it
+            // as a signed value.
+            let mask = ((1u64 << 63) as i64) >> unsigned_value.leading_zeros();
+            (unsigned_value as i64) | mask
+        };
+
+        Ok((
+            FlexInt::new(flex_uint.size_in_bytes(), signed_value),
+            remaining,
+        ))
+    }
+
+    /// Reads a [`FlexUInt`] from the buffer.
+    #[inline]
+    pub fn read_flex_uint(self) -> ParseResult<'a, FlexUInt> {
```

Comment:
minor organizational suggestion—it took me a little bit to figure out why `FlexUInt` needed data or a constructor, since the `write_u64` function doesn't use instances of `FlexUInt` at all. Consider extracting the decoding logic into `FlexUInt` so that we can keep related things close together.

E.g.: add `read(&[u8]) -> SomeResult<FlexUInt>` to `FlexUInt`, and then this function can be reduced to:
```rust
pub fn read_flex_uint(self) -> ParseResult<'a, FlexUInt> {
    let flex_uint = FlexUInt::read(self.bytes())?;
    let remaining = self.consume(flex_uint.size_in_bytes());
    Ok((flex_uint, remaining))
}
```

Reply:
Done in commit c572b6d7c81a2ccdfd06f5d025aa2effa120d8f1.